# Comparing Dense vs. SVD Retrieval
**Author:** Hai Ran Zheng

**Date:** March 14th, 2025

**Description:** 
Compare train time, inference time, and quality of SVD and Dense Embedding Retrieval

# Import

In [1]:
import pandas as pd
import numpy as np
import re
import os
from nltk import word_tokenize
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.model_selection import train_test_split
import nltk
from typing import Literal, Union

# check to see if the nltk data has been downloaded in the virtual environment
if not os.path.exists(os.path.join(os.path.expanduser("~"), "nltk_data")):
    # download the nltk data
    nltk.download('stopwords')
    nltk.download('punkt_tab')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\harri\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\harri\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


# Utilities

In [4]:
def left_align(df: pd.DataFrame):
    left_aligned_df = df.style.set_properties(**{'text-align': 'left'})
    left_aligned_df = left_aligned_df.set_table_styles(
        [dict(selector='th', props=[('text-align', 'left')])]
    )
    return left_aligned_df

In [5]:
pd.set_option('max_colwidth', 400)

# List of Queries For Testing
<div style="font-size:20px;font-weight:700">
    Direct Vehicle Issues
</div>

- Speaker rattles
- Clicking sound steering left and right
- Brake lamps stay on
- Air bag issues from faulty sensor
- Bad smell can be observed due to break pads overheating.
- Overheating can be observed due to power train issues.
- Door open/close issues can be observed due to door latch issue, child safety mode malfunctions.
- Harsh clutch movements can be observed due to transmission/powertrain issues.
- Steering wheel issues can be observed due to unbalanced tires.
- Seat adjustment issues can be observed due to debri in seat latch.
- Light issues can be observed due to faulty electrical sensors.

 - **Latch** is an interesting one, as well as **axle**, since it could refer to different parts of a vehicle.


In [9]:
PARENT_DIR = os.getcwd().rsplit(os.sep, maxsplit=1)[0]
DATASET_DIR = os.path.join(PARENT_DIR, "Datasets")

In [10]:
rand_path = f"{DATASET_DIR}/complaints_and_recalls.csv"

In [11]:
os.path.normpath(rand_path)

'c:\\Users\\harri\\OneDrive\\Documents\\Programming\\UMich_Data_Science\\SIADs_Audio_Text_SRS\\Datasets\\complaints_and_recalls.csv'

In [12]:
df = pd.read_csv(f"{DATASET_DIR}/complaints_and_recalls.csv")

In [13]:
df["IS_COMPLAINT"].value_counts()

IS_COMPLAINT
False    26153
True     11615
Name: count, dtype: int64

# SVD Retrieval

1. Convert Documents into TF-IDF vector
2. Train Scikit-Learn's SVD algorithm on train set
3. Retrieve relevant documents given query
4. Show relevant topics

In [15]:
# create a streamlit app that allows users to input a text query and get the top 5 most similar documents from the corpus using the TextClassifier class from import os
import os
import pickle

class TextClassifier:
    def __init__(self, dataset_path, column_name:str, n_dimensions:int):
        # set the random state for reproducibility
        self.random_state = 42
        # create a TfidfVectorizer object
        self.vectorizer = TfidfVectorizer()
        # set the number of dimensions to reduce the vectorized data to
        self.num_dimensions = n_dimensions
        self.dataset_path = os.path.abspath(dataset_path) # normalize
        self.dataset_dir, self.dataset_basename = self.dataset_path.rsplit(
            os.sep, maxsplit=1
        )
        self.column_name = column_name
        self.column_name_cleaned = column_name + "_CLEANED"
        # create variables to store the training, test, and validation sets for class functions
        self.column_name_cleaned = None
        self.df_train = None
        self.df_test = None
        self.df_validation = None
        self.x_train_vect = None
        self.x_test_vect = None
        self.x_validation_vect = None
        self.lsa = None
        self.vectorized_train = None
        self.vectorized_test = None
        self.vectorized_validation = None

    def read_dataset(self):
        """Read the dataset from the csv file"""
        self.df = pd.read_csv(self.dataset_path)

    # create a function that will be called from a apply lambda function to process the text from a dataframe with the "CDESCR" column
    def process_text(text):
        """
        Process text by tokenizing, removing stop words, and stemming
        """
        #print(text)
        stop_words = set(stopwords.words("english"))
        stemmer = PorterStemmer()
        
        # make sure the text is not empty and or nan
        if not text or pd.isna(text):
            return []

        # tokenize the text and remove stop words and stem the words
        content_cleaned = [stemmer.stem(word) for word in word_tokenize(text.lower()) if word not in stop_words]

        # remove the punctuation from the content_cleaned list
        content_cleaned = [word for word in content_cleaned if word.isalnum()]
        
        return content_cleaned

    def load_pickle_if_paths_exist(BASE_PATH, basenames_to_check : list):
        pickles = []
        for basename_to_check in basenames_to_check:
            if os.path.exists(f"{BASE_PATH}/{basename_to_check}"):
                with open(f"{BASE_PATH}/{basename_to_check}", "rb") as f:
                    pickles.append(pickle.load(f))

        return pickles
    
    def process_dataframe(self, train_size=0.7, test_size=0.2, validation_size=0.1, subgroup=""):
        """
        Process the text in the "CDESCR" column and create a new column "CDESCR_CLEANED" with the processed text
        """
        basename = f"{self.column_name}_{self.num_dimensions}"
        # get current working directory
        cwd = os.getcwd()
        # change the Example folder to Datasets folder in the cwd path
        desired_save_path = os.path.join(self.dataset_dir, subgroup)
        # desired_save_path = cwd + "\\Datasets"
        # create a folder path to save the pickle files
        if not os.path.exists(desired_save_path):
            os.makedirs(desired_save_path)

        # check to see if there is a pickle file for the dataframe with the processed text
        if os.path.exists(desired_save_path + '//' + basename + "_df.pkl"):
            # load the pickle file
            with open(desired_save_path + '//' + basename + "_df.pkl", "rb") as f:
                self.df = pickle.load(f)
        else:
            # process the text in the "CDESCR" column and create a new column "CDESCR_CLEANED" with the processed text
            self.df[self.column_name_cleaned] = self.df[self.column_name].apply(lambda x: TextClassifier.process_text(x))
            # create a pickle file for the dataframe with the processed text
            with open(desired_save_path + '//' + basename + "_df.pkl", "wb") as f:
                pickle.dump(self.df, f)

        # split the df_complaints dataframe into a test, train, and validation set with a 70/20/10 split
        self.df_train, self.df_test = train_test_split(self.df, test_size=(1-train_size), random_state=self.random_state)
        self.df_test, self.df_validation = train_test_split(self.df_test, test_size=(validation_size/(1-train_size)), random_state=self.random_state)

        # check to see if there is a pickle file for the vectorizer
        if (
            os.path.exists(desired_save_path + '//' + basename + "_vectorizer.pkl") 
            and os.path.exists(desired_save_path + '//' + basename + "_x_train_vect.pkl") 
            and os.path.exists(desired_save_path + '//' + basename + "_x_test_vect.pkl") 
            and os.path.exists(desired_save_path + '//' + basename + "_x_validation_vect.pkl")
        ):
            # load the pickle file
            with open(desired_save_path + '//' + basename + "_vectorizer.pkl", "rb") as f:
                self.vectorizer = pickle.load(f)
            with open(desired_save_path + '//' + basename + "_x_train_vect.pkl", "rb") as f:
                self.x_train_vect = pickle.load(f)
            with open(desired_save_path + '//' + basename + "_x_test_vect.pkl", "rb") as f:
                self.x_test_vect = pickle.load(f)
            with open(desired_save_path + '//' + basename + "_x_validation_vect.pkl", "rb") as f:
                self.x_validation_vect = pickle.load(f)
        else:
            # fit the vectorizer on the "CDESCR_CLEANED" column and transform the "CDESCR_CLEANED" column into a vectorized format
            # apply a lambda function to join the list of words in the "CDESCR_CLEANED" column into a string
            self.x_train_vect = self.vectorizer.fit_transform(self.df_train[self.column_name_cleaned].apply(lambda x: " ".join(x)))
            self.x_test_vect = self.vectorizer.transform(self.df_test[self.column_name_cleaned].apply(lambda x: " ".join(x)))
            self.x_validation_vect = self.vectorizer.transform(self.df_validation[self.column_name_cleaned].apply(lambda x: " ".join(x)))        
            #print(self.x_train_vect.shape, self.x_test_vect.shape, self.x_validation_vect.shape)
            # create a pickle file for the vectorizer
            with open(desired_save_path + '//' + basename + "_vectorizer.pkl", "wb") as f:
                pickle.dump(self.vectorizer, f)
            with open(desired_save_path + '//' + basename + "_x_train_vect.pkl", "wb") as f:
                pickle.dump(self.x_train_vect, f)
            with open(desired_save_path + '//' + basename + "_x_test_vect.pkl", "wb") as f:
                pickle.dump(self.x_test_vect, f)
            with open(desired_save_path + '//' + basename + "_x_validation_vect.pkl", "wb") as f:
                pickle.dump(self.x_validation_vect, f)

        # check to see if there is a pickle file for the lsa
        if os.path.exists(desired_save_path + '//' + basename + "_lsa.pkl") and os.path.exists(desired_save_path + '//' + basename + "_complaints_vectorized_train.pkl") and os.path.exists(desired_save_path + '//' + basename + "_complaints_vectorized_test.pkl") and os.path.exists(desired_save_path + '//' + basename + "_complaints_vectorized_validation.pkl"):
            # load the pickle file
            with open(desired_save_path + '//' + basename + "_lsa.pkl", "rb") as f:
                self.lsa = pickle.load(f)
            with open(desired_save_path + '//' + basename + "_complaints_vectorized_train.pkl", "rb") as f:
                self.complaints_vectorized_train = pickle.load(f)
            with open(desired_save_path + '//' + basename + "_complaints_vectorized_test.pkl", "rb") as f:
                self.complaints_vectorized_test = pickle.load(f)
            with open(desired_save_path + '//' + basename + "_complaints_vectorized_validation.pkl", "rb") as f:
                self.complaints_vectorized_validation = pickle.load(f)
        else:
            # perform LSA on the vectorized data to reduce the dimensionality
            self.lsa = TruncatedSVD(n_components=self.num_dimensions, random_state=self.random_state)
            self.complaints_vectorized_train = self.lsa.fit_transform(self.x_train_vect)
            self.complaints_vectorized_test = self.lsa.transform(self.x_test_vect)
            self.complaints_vectorized_validation = self.lsa.transform(self.x_validation_vect)
            
            # create a pickle file for the vectorized training data
            with open(desired_save_path + '//' + self.column_name + "_lsa.pkl", "wb") as f:
                pickle.dump(self.lsa, f)
            with open(desired_save_path + '//' + self.column_name + "_complaints_vectorized_train.pkl", "wb") as f:
                pickle.dump(self.complaints_vectorized_train, f)
            with open(desired_save_path + '//' + self.column_name + "_complaints_vectorized_test.pkl", "wb") as f:
                pickle.dump(self.complaints_vectorized_test, f)
            with open(desired_save_path + '//' + self.column_name + "_complaints_vectorized_validation.pkl", "wb") as f:
                pickle.dump(self.complaints_vectorized_validation, f)

        return self.df, self.df_train, self.df_test, self.df_validation, self.lsa, self.vectorizer
    
    # A functionalize the process of finding similar complaints to a query text
    def find_similar_complaint(self, query_text:str):
        """
        Find the most similar complaint to a query text in the training set
        """
        # process the query text
        query_text_cleaned = TextClassifier.process_text(query_text)
        # vectorize the query text
        query_vectorized = self.vectorizer.transform([" ".join(query_text_cleaned)])
        # reduce the dimensionality of the query vector
        query_vectorized_lsa = self.lsa.transform(query_vectorized)
        # find the cosine similarity between the query vector and all the complaints in the training set
        cosine_similarities_query = cosine_similarity(query_vectorized_lsa, self.complaints_vectorized_train)
        # get the index of the most similar complaint in the training set
        most_similar_index_query = cosine_similarities_query.argmax()
        print(most_similar_index_query)
        # get the most similar complaint in the training set
        most_similar_complaint_train_query = self.df_train.iloc[most_similar_index_query]
        return most_similar_complaint_train_query
    
    def find_similar(self, query_text:str, 
                                    dataset_to_check : Literal[
                                        "train", 
                                        "validation",
                                        "test"    
                                    ] = "train",
                                    similarity_fn=cosine_similarity,
                                    top=5):
        """
        Find the most similar complaint to a query text in the training set
        """
        # process the query text
        query_text_cleaned = TextClassifier.process_text(query_text)
        # vectorize the query text
        query_vectorized = self.vectorizer.transform([" ".join(query_text_cleaned)])
        # reduce the dimensionality of the query vector
        query_vectorized_lsa = self.lsa.transform(query_vectorized)
        # find the cosine similarity between the query vector and all the complaints in the training set
        
        if dataset_to_check == "train":
            cosine_similarities_query = similarity_fn(query_vectorized_lsa, self.complaints_vectorized_train)
        elif dataset_to_check == "validation":
            cosine_similarities_query = similarity_fn(query_vectorized_lsa, self.complaints_vectorized_validation)
        elif dataset_to_check == "test":
            cosine_similarities_query = similarity_fn(query_vectorized_lsa, self.complaints_vectorized_test)
    
        # get the index of the most similar complaints
        most_similar_indices_query = cosine_similarities_query.squeeze().argsort()[:-top-1:-1]

        # get the most similar complaints
        most_similar_elements = self.df_train.iloc[most_similar_indices_query]
        most_similar_elements["cosine_similarities"] = cosine_similarities_query.squeeze()[most_similar_indices_query]
        return most_similar_elements
    
    def run_training_pipeline(self, subgroup=""):
        # Load dataset in self.df
        self.read_dataset()

        # Process DataFrame and run SVD
        self.process_dataframe(subgroup=subgroup)

# run the below code if main script
# if __name__ == "__main__":
#     # https://www.nhtsa.gov/nhtsa-datasets-and-apis#recalls
#     # read in C:\Repo\SIADs_Audio_Text_SRS\Example\COMPLAINTS_RECEIVED_2025-2025.txt into a pandas dataframe, where the columns are RCL
#     df_complaints = pd.read_csv("C:\\Repo\\SIADs_Audio_Text_SRS\\Datasets\\COMPLAINTS_RECEIVED_2025-2025.txt", sep='\t', header=None, index_col=0)
#     df_complaints.columns = ['ODINO', 'MFR_NAME', 'MAKETXT', 'MODELTXT', 'YEARTXT', 'CRASH', 'FAILDATE', 'FIRE', 'INJURED', 'DEATHS', 'COMPDESC', 'CITY', 'STATE', 'VIN', 'DATEA', 'LDATE', 'MILES', 'OCCURENCES', 'CDESCR', 'CMPL_TYPE', 'POLICE_RPT_YN', 'PURCH_DT', 'ORIG_OWNER_YN', 'ANTI_BRAKES_YN', 'CRUISE_CONT_YN', 'NUM_CYLS', 'DRIVE_TRAIN', 'FUEL_SYS', 'FUEL_TYPE',
#                 'TRANS_TYPE', 'VEH_SPEED', 'DOT', 'TIRE_SIZE', 'LOC_OF_TIRE', 'TIRE_FAIL_TYPE', 'ORIG_EQUIP_YN', 'MANUF_DT', 'SEAT_TYPE', 'RESTRAINT_TYPE', 'DEALER_NAME', 'DEALER_TEL', 'DEALER_CITY', 'DEALER_STATE', 'DEALER_ZIP', 'PROD_TYPE', 'REPAIRED_YN', 'MEDICAL_ATTN', 'VEHICLES_TOWED_YN']

#     # create a list of unique manufacturers in the "MFR_NAME" column
#     list_of_manufacturers = df_complaints["MFR_NAME"].unique()

#     # testing out the functionilzed code

#     # call the TextClassifier class and create an instance of it as text_classifier
#     # pass in the df_complaints dataframe and the "CDESCR" column
#     text_classifier = TextClassifier(df_complaints, "CDESCR")
#     # process the text in the "CDESCR" column
#     text_classifier.process_dataframe()

#     # use one of the complaints in the test set as a query to find the most similar complaint in the training set
#     #complaint_test_query = text_classifier.df_test["CDESCR"].iloc[5]
#     #complaint_test_query = text_classifier.df_test["CDESCR"].iloc[4]
#     #complaint_test_query = "Car won't start and makes a clicking noise"
#     complaint_test_query = "Battery dies after a few days of not driving the car"

#     print(complaint_test_query)
#     # find the most similar complaint to the complaint test
#     most_similar_complaint = text_classifier.find_similar_complaint(complaint_test_query)
#     # print the most similar complaint with the below columns
#     print(most_similar_complaint[["ODINO", "MFR_NAME", "MAKETXT", "MODELTXT", "YEARTXT", "CDESCR"]])
#     print(most_similar_complaint["CDESCR"])

In [16]:
clf_svd = TextClassifier(f'{DATASET_DIR}/complaints_and_recalls.csv', column_name='CDESCR',n_dimensions=384)

In [17]:
_ = clf_svd.run_training_pipeline(subgroup="SVD_384")

In [18]:
len(clf_svd.df_train)

26444

## Training Time

2 minute 10 second training time for 26444 documents. That's about 26644/130 = 205 documents per second on my computer. Here are my computer specs:

In [17]:
!cpuinfo

Python Version: 3.12.7.final.0 (64 bit)
Cpuinfo Version: 9.0.0
Vendor ID Raw: AuthenticAMD
Hardware Raw: 
Brand Raw: AMD Ryzen 5 8640HS w/ Radeon 760M Graphics
Hz Advertised Friendly: 3.4930 GHz
Hz Actual Friendly: 2.4860 GHz
Hz Advertised: (3493000000, 0)
Hz Actual: (2486000000, 0)
Arch: X86_64
Bits: 64
Count: 12
Arch String Raw: AMD64
L1 Data Cache Size: 
L1 Instruction Cache Size: 
L2 Cache Size: 6291456
L2 Cache Line Size: 1024
L2 Cache Associativity: 6
L3 Cache Size: 16777216
Stepping: 2
Model: 117
Family: 25
Processor Type: 
Flags: 3dnow, 3dnowext, 3dnowprefetch, abm, adx, aes, apic, avx, avx2, avx512bitalg, avx512bw, avx512cd, avx512dq, avx512f, avx512ifma, avx512vbmi, avx512vbmi2, avx512vl, avx512vnni, avx512vpopcntdq, bmi1, bmi2, clflush, clflushopt, clwb, cmov, cmp_legacy, cr8_legacy, cx16, cx8, de, dts, erms, f16c, fma, fpu, fxsr, gfni, ht, hypervisor, ia64, invpcid, lahf_lm, mca, mce, misalignsse, mmx, monitor, movbe, msr, mtrr, osvw, osxsave, pae, pat, pclmulqdq, perfctr_c

Ignore python version as it is my global python version. I'm using a venv that has a different python version (3.10.16)

## Inference Time + Quality of Retrieval

In [27]:
columns_to_inspect = ["CDESCR", "cosine_similarities", "IS_COMPLAINT"]

In [18]:
similar_elements = clf_svd.find_similar(
    query_text="Speaker rattles",
    dataset_to_check="train",
    top=50
)

C:\Users\harri\AppData\Local\Temp\ipykernel_37020\578085221.py:209: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  most_similar_elements["cosine_similarities"] = cosine_similarities_query.squeeze()[most_similar_indices_query]


In [ ]:
c_descr = left_align(similar_elements[columns_to_inspect].head(10))

In [20]:
pd.set_option('max_colwidth', 400)

In [21]:
c_descr

In [22]:
similar_elements = clf_svd.find_similar(
    query_text="Clicking sounds steering left and right",
    dataset_to_check="train",
    top=50
)

C:\Users\harri\AppData\Local\Temp\ipykernel_37020\578085221.py:209: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  most_similar_elements["cosine_similarities"] = cosine_similarities_query.squeeze()[most_similar_indices_query]


In [ ]:
c_descr = left_align(similar_elements[columns_to_inspect].head(10))

In [24]:
c_descr

In [25]:
similar_elements = clf_svd.find_similar(
    query_text="brake lamp stay on",
    dataset_to_check="train",
    top=50
)

C:\Users\harri\AppData\Local\Temp\ipykernel_37020\578085221.py:209: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  most_similar_elements["cosine_similarities"] = cosine_similarities_query.squeeze()[most_similar_indices_query]


In [28]:
left_align(similar_elements[columns_to_inspect].head(10))

In [32]:
similar_elements = clf_svd.find_similar(
    query_text="air bag",
    dataset_to_check="train",
    top=50
)

C:\Users\harri\AppData\Local\Temp\ipykernel_37020\578085221.py:209: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  most_similar_elements["cosine_similarities"] = cosine_similarities_query.squeeze()[most_similar_indices_query]


In [34]:
similar_elements["IS_COMPLAINT"].value_counts()

IS_COMPLAINT
False    47
True      3
Name: count, dtype: int64

In [35]:
left_align(similar_elements[columns_to_inspect].head(10))

In [53]:
%%time
similar_elements = clf_svd.find_similar(
    query_text="bad smell in car. Rubber",
    dataset_to_check="train",
    top=50
)

CPU times: total: 125 ms
Wall time: 123 ms


C:\Users\harri\AppData\Local\Temp\ipykernel_37020\578085221.py:209: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  most_similar_elements["cosine_similarities"] = cosine_similarities_query.squeeze()[most_similar_indices_query]


In [37]:
left_align(similar_elements[columns_to_inspect].head(10))

This does not make sense at all.

In [52]:
%%time
similar_elements = clf_svd.find_similar(
    query_text="Door has trouble opening and closing",
    dataset_to_check="train",
    top=50
)

CPU times: total: 109 ms
Wall time: 116 ms


C:\Users\harri\AppData\Local\Temp\ipykernel_37020\578085221.py:209: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  most_similar_elements["cosine_similarities"] = cosine_similarities_query.squeeze()[most_similar_indices_query]


In [39]:
left_align(similar_elements[columns_to_inspect].head(10))

In [51]:
%%time
similar_elements = clf_svd.find_similar(
    query_text="Seat latch",
    dataset_to_check="train",
    top=50
)

CPU times: total: 250 ms
Wall time: 410 ms


C:\Users\harri\AppData\Local\Temp\ipykernel_37020\578085221.py:209: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  most_similar_elements["cosine_similarities"] = cosine_similarities_query.squeeze()[most_similar_indices_query]


In [50]:
left_align(similar_elements[columns_to_inspect].head(10))

Both seat belt and seat latches are queried. 

## Grade

Overall grade:
1. Excellent Training Time even on CPU: ~205 documents per Second
2. Fast Inference: ~150ms. For 1 million documents, this is not very fast (~41 hours)
3. Some topics not covered (e.g. does not understand what 'bad smell' or 'overheating' is), but it is able to capture semantic relationships (query "speaker rattles" does uncover car radio problems).

Another problem is that complaints information is more specific to a vehicle issue than recall information, which may mention more things such as class of vehicle. 

# Dense Embedding Retrieval

1. Convert documents into vector embeddings (e.g. All-MiniLM for general purpose embeddings, or use Roberta for more semantic embeddings)
2. Split documents into chunks and store in vector index/database. 
3. Send query to vector database

In [19]:
from sentence_transformers import SentenceTransformer
import torch

In [45]:
class HuggingFaceEmbeddings:
    def __init__(self, dataset_path, model_name, column_name,
                 similarity_fn_name : Union[
                            "cosine", "dot"
                ] = "cosine"):
        self.column_name = column_name
        self.model_name = model_name
        self.dataset_path = dataset_path
        self.df = pd.read_csv(os.path.join(self.dataset_path, "complaints_and_recalls.csv"))
        self.model_path = os.path.join(self.dataset_path, "huggingface", 
                                      self.column_name)

        # Load sentence transformer if it has already been saved
        if os.path.exists(os.path.join(self.model_path, "model")):
            self.model = SentenceTransformer(os.path.join(self.model_path, "model"),
                                             similarity_fn_name=similarity_fn_name)
        else:
            self.model = SentenceTransformer(self.model_name,
                                             similarity_fn_name=similarity_fn_name)

    def split_text_by_period(text, max_length=200):
        '''
            Chunking not in use currently.
        '''
        sentences = re.split(r'\.\s+', text)  # Split by period + space
        chunks = []
        current_chunk = ""

        for sentence in sentences:
            if len(current_chunk) + len(sentence) < max_length:
                current_chunk += sentence + ". "
            else:
                current_chunk += sentence + "."
                chunks.append(current_chunk.strip())
                current_chunk = "" # Reset current chunk

        if current_chunk:
            chunks.append(current_chunk.strip())

        return chunks

    def dataset_processing(self):
        '''
            Not Used Yet
        '''
        self.df[self.column_name + "_CLEANED"] = (
            self.df[self.column_name]
            .apply(lambda x: self.split_text_by_period(x))
        )
        self.df = self.df[
            ["index", self.column_name + "_CLEANED"]
        ].explode(self.column_name + "_CLEANED")

    def encode_embeddings(self):
        '''
        TODO: Consider GPU acceleration
        '''
        
        # Put preprocessing Here

        # Encoding here
        if not os.path.exists(self.model_path):
            os.makedirs(self.model_path, exists_ok=True)
            # Load model, dataframe, and encode text.
            self.embeddings = self.model.encode(self.df[self.column_name], show_progress_bar=True)
            with open(os.path.join(self.model_path, f"{self.model_name}-embeddings.pkl"), "wb") as f:
                pickle.dump(self.embeddings, f)
        else:
            print("Embeddings found. Loading embeddings.")
            with open(os.path.join(self.model_path, f"{self.model_name}-embeddings.pkl"), "rb") as f:
                self.embeddings = pickle.load(f)
    
    def query_embeddings(self, query, top=20):
        query_embedding = self.model.encode(query)
        model_similarities = self.model.similarity(
            query_embedding,
            self.embeddings
        )
        values, indices = torch.topk(model_similarities, largest=True, k=top, dim=1)
        return self.df.iloc[indices[0].tolist()], values

    def run_training_pipeline(self):
        '''
            Technically no training is done.
            Strictly speaking, this is simply a processing pipeline
        '''
        self.encode_embeddings()



In [46]:
hg = HuggingFaceEmbeddings(
    dataset_path=DATASET_DIR,
    model_name="all-MiniLM-L6-v2",
    column_name="CDESCR",
    similarity_fn_name="cosine"
)

In [47]:
hg.run_training_pipeline()

Embeddings found. Loading embeddings.


In [48]:
returned_docs, sims = hg.query_embeddings(
    query="Car won't start and makes a clicking noise",
    top=50
)

In [49]:
returned_docs, sims = hg.query_embeddings(
    query="Car won't start and makes a clicking noise",
    top=50
)

In [50]:
pd.set_option('max_colwidth', 200)

In [51]:
returned_docs.head(5)

index  \
1213    1213   
10951  10951   
5952    5952   
5953    5953   
421      421   

                                                                                                                                                                                                        CDESCR  \
1213   Coolant control valve failure (P1098) after 50k service. When this engine code appears the vehicle goes into reduced power mode. While it does this at inopportune time and creates potential for co...   
10951  Whenever the car engine isn't warm and I slam on the brakes, they barely respond. I have replaced the master cylinder, abs hydraulic module, brake booster, calipers, pads, and rotors, and it still...   
5952     The catalytic converter had to be replaced after 3000 miles. Vehicle was purchased in September and it had to be replaced in December. Had to take vehicle in twice for them to diagnose the problem.   
5953   The center console and the screen behind the sharing wheel black at the same time while I was driving. I couldn't see the speed I was going while driving. I tried turning off the car, taking the k...   
421    About 2 weeks ago started hearing what sounded like a earing going bearing from under the hood on driver side. I took it to the dealership yesterday, told them im pretty sure bearing going out on ...   

                        COMPDESC                    MMYTXT  RECORDID  \
1213                 POWER TRAIN      GMC SIERRA 1500 2022  11644914   
10951             SERVICE BRAKES        MERCURY MILAN 2010  11639152   
5952            UNKNOWN OR OTHER  JEEP GRAND CHEROKEE 2024  11643050   
5953            UNKNOWN OR OTHER       HONDA PROLOGUE 2024  11634241   
421    ENGINE, ELECTRICAL SYSTEM             RAM 1500 2024  11643293   

       NUMCOMPLAINTS  IS_COMPLAINT  
1213             1.0          True  
10951            1.0          True  
5952             1.0          True  
5953             1.0          True  
421              2.0          True

In [ ]:
returned_docs["IS_COMPLAINT"].value_counts()

IS_COMPLAINT
True    50
Name: count, dtype: int64

In [ ]:
returned_docs["CDESCR"].iloc[4]

'A few months after I purchased the car in July, I started hearing a clicking noise when turning from a full stop. In December I was told it was axel issue. It’s not yet been resolved. I’m still driving the car. '

In [ ]:
returned_docs, sims = hg.query_embeddings(
    query="Car won't start and makes a clicking noise",
    top=50
)

In [ ]:
returned_docs, sims = hg.query_embeddings(
    query="Bad smell overheating",
    top=50
)

In [ ]:
sims

tensor([[0.5880, 0.5559, 0.5389, 0.5370, 0.5323, 0.5210, 0.5096, 0.4913, 0.4862,
         0.4762, 0.4754, 0.4722, 0.4673, 0.4656, 0.4648, 0.4638, 0.4607, 0.4595,
         0.4584, 0.4577, 0.4542, 0.4540, 0.4478, 0.4463, 0.4390, 0.4371, 0.4361,
         0.4347, 0.4339, 0.4301, 0.4287, 0.4286, 0.4279, 0.4247, 0.4242, 0.4230,
         0.4216, 0.4207, 0.4191, 0.4187, 0.4170, 0.4132, 0.4111, 0.4100, 0.4100,
         0.4089, 0.4074, 0.4073, 0.4067, 0.4064]])

In [ ]:
returned_docs.head(5)

index  \
2057    2057   
4000    4000   
1314    1314   
2848    2848   
10873  10873   

                                                                                                                                                                                                                                                                                                                                                                                                                CDESCR  \
2057                                                                                                                                                                                                                                                                                                                                                                                Front heater fan smelling of smoke   
4000   My Jeep eats up thermostats too often. I have had it for 3 years and had it replaced 4 times. The last time , one year ago, it took the repair shop a week to finally install one that would hold. It needs it again. I also smell something burning when I accelerate onto the freeway since owning it. My engines fan runs high aIl the time also. I read many complaints online about people having overh...   
1314                                                                                                                                                                                                                                                                                                                                                                Coolant is leaking and causing engine to overheat.   
2848                                                                                                                                         I just bought this used car and just drove around town. It has 101,724 miles and it smells like something burning when I start it up and when I drive it. I know it's coming from somewhere under the hood. I'm paying 9500.00 for this car and already have burnt smell.   
10873                                                                                                                                                                                                                                                    When my heat is on there is a burning smell when the heat on and my right side passages door will not allow me to open the door from the inside or out sides    

                COMPDESC                    MMYTXT  RECORDID  NUMCOMPLAINTS  \
2057   ELECTRICAL SYSTEM           LION LIONC 2024  11642408            1.0   
4000              ENGINE  JEEP GRAND CHEROKEE 2014  11635444            1.0   
1314              ENGINE       CHEVROLET TRAX 2017  11640557            1.0   
2848    UNKNOWN OR OTHER        JEEP CHEROKEE 2016  11640183            1.0   
10873   UNKNOWN OR OTHER       HYUNDAI SONATA 2013  11637091            1.0   

       IS_COMPLAINT  
2057           True  
4000           True  
1314           True  
2848           True  
10873          True

In [ ]:
returned_docs, sims = hg.query_embeddings(
    query="Bad smell overheating",
    top=50
)

In [ ]:
returned_docs, sims = hg.query_embeddings(
    query="Door opens even when child safety mode is activated",
    top=50
)

In [ ]:
returned_docs.head(5)

index  \
10937  10937   
1206    1206   
1205    1205   
3733    3733   
5846    5846   

                                                                                                                                                                                                                                                                                                                                                                                                                CDESCR  \
10937  When turning on the child safety locks, the doors repeat a clicking sound for a few minutes sounding as if the doors are locking. After the noise stops, the dashboard shows the message "child safety lock failure". My small child has been winding the window down since I cannot lock the windows and doors properly. I am afraid she will get a body part caught in the window or open the door while d...   
1206                                                                                                                                                                                            Child safety locks failed and clicked attempting to engage until turned off.With the locks unable to engage, children able to utilize windows and open doors which put their safety at risk. Replaced under warranty.    
1205   Child Safety Lock Failure notice - both parked or driving, I will hear a clicking noise coming from the doors. Sometimes I will receive a Child Safety Lock Failure error message on the dash. This would sporadically continue to happen while driving and even if my vehicle is parked and off and I happen to do something like open the doors. The vehicle has been serviced numerous times by Hyundai d...   
3733                                                                                                                    Loaded children into second row of vehicle, confirmed all doors closed. Began driving, passenger rear door swung open. Child DID NOT engage door handle. Pulled over to attempt to close door. Door latch would not engage, could not close or lock door. Still unable to close and lock door.   
5846   The back doors will no longer open when unlocked. This poses a safety risk. I was unable to get my child out of his car seat when the locks or locking mechanism went out. I had to climb into the back seat from the front. The door do not open either from the interior or exterior of the vehicle. A mechanic stated that this could be from faulty solenoids. I was unable to get a print out of a quot...   

                                 COMPDESC                 MMYTXT  RECORDID  \
10937                   ELECTRICAL SYSTEM  HYUNDAI PALISADE 2021  11642064   
1206                    ELECTRICAL SYSTEM  HYUNDAI PALISADE 2021  11643828   
1205                    ELECTRICAL SYSTEM  HYUNDAI PALISADE 2021  11634797   
3733   LATCHES/LOCKS/LINKAGES:DOORS:LATCH        VOLVO XC90 2016  11640401   
5846     UNKNOWN OR OTHER, STRUCTURE:BODY    HYUNDAI SONATA 2013  11634770   

       NUMCOMPLAINTS  IS_COMPLAINT  
10937            1.0          True  
1206             1.0          True  
1205             1.0          True  
3733             1.0          True  
5846             2.0          True

In [ ]:
returned_docs, sims = hg.query_embeddings(
    query="Mercedes Benz issue with fuel leakage",
    top=50
)

In [ ]:
returned_docs.head(5)

index  \
2110    2110   
2089    2089   
9310    9310   
27123  15499   
18446   6822   

                                                                                                                                                                                                                                                                                                                                                                                                                CDESCR  \
2110   Fuel tank is leaking at connection between fill tube and tank. Multiple recalls have been issued apparently for similar problems with cars of other years. With less than 1/2 tank, the fuel leak is not as pronounced. With the tank full, the tank drips fuel. Fire department in Riverside was called and confirmed issue. Upon bringing vehicle to Clearwater Mercedes and other mechanics, the problem ...   
2089                                                                                                                                                                                                                                                                                                                                                                                         Fuel leaking from engine    
9310                                                                                                                             The seal that connects the fuel level sensor to the fuel tank has failed, allowing fuel to leak onto the top of the tank when it’s filled. This creates a serious risk of fire or explosion. Unfortunately, Mercedes does not offer a remedy unless you pay a prohibitively high fee.   
27123  ON CERTAIN SPORT UTILITY VEHICLES, THE FUEL RETURN HOSE BETWEEN FUEL TANK AND RETURN LINE CAN BE PERMEATED WITH FUEL. THIS SITUATION RESULTS FROM THE EXPANSION AND CONTRACTION OF THE HOSE UNDER PRESSURE CAUSING MICROCRACKS IN THE HOSE MATERIAL.\n\n THESE MICROCRACKS MAY CAUSE FUEL LEAKAGE. FUEL LEAKAGE IN THE PRESENCE OF AN IGNITION SOURCE COULD RESULT IN A FIRE.\n\n DEALERS WILL INSPECT AND R...   
18446                                                                                                                                                                                                        FUEL LEAKAGE MAY OCCUR IN THE ENGINE BAY AT FUEL LINE CONNECTIONS OR AT THE FUEL PUMP.\n\n \n\n THE DEALER WILL INSPECT THE FUEL LINES AND FUEL PUMP AND WILL REPLACE THEM, IF NECESSARY, WITHOUT CHARGE.   

                                                               COMPDESC  \
2110                                             FUEL/PROPULSION SYSTEM   
2089                                     ENGINE, FUEL/PROPULSION SYSTEM   
9310                                             FUEL/PROPULSION SYSTEM   
27123  FUEL SYSTEM, GASOLINE:DELIVERY:HOSES, LINES/PIPING, AND FITTINGS   
18446                                    FUEL SYSTEM, GASOLINE:DELIVERY   

                                                                                                                                         MMYTXT  \
2110                                                                                                                    MERCEDES BENZ C300 2013   
2089                                                                                                                      CHEVROLET MALIBU 2017   
9310                                                                                                                    MERCEDES BENZ E550 2008   
27123                                                                                            MERCEDES BENZ G55 2004, MERCEDES BENZ G55 2005   
18446  TRIUMPH SPITFIRE 1974, TRIUMPH SPITFIRE 1971, TRIUMPH SPITFIRE 1976, TRIUMPH SPITFIRE 1972, TRIUMPH SPITFIRE 1973, TRIUMPH SPITFIRE 1975   

                                       RECORDID  NUMCOMPLAINTS  IS_COMPLAINT  
2110                                

# Try to Improve Results

## SVD

## Dense